File: go/store_postgres_test.go
//go:build integration

package fasten

import (
	"context"
	"database/sql"
	"fmt"
	"os"
	"testing"
	"time"

	_ "github.com/lib/pq"
)

func pgDB(t *testing.T) *sql.DB {
	t.Helper()
	dsn := os.Getenv("FASTEN_TEST_POSTGRES_DSN")
	if dsn == "" {
		t.Skip("FASTEN_TEST_POSTGRES_DSN not set")
	}
	db, err := sql.Open("postgres", dsn)
	if err != nil {
		t.Fatalf("sql.Open: %v", err)
	}
	if err := db.Ping(); err != nil {
		db.Close()
		t.Fatalf("db.Ping: %v", err)
	}
	t.Cleanup(func() { db.Close() })
	return db
}

func uniqueTable(prefix string) string {
	return fmt.Sprintf("%s_%d", prefix, time.Now().UnixNano())
}

func dropTable(t *testing.T, db *sql.DB, table string) {
	t.Helper()
	if _, err := db.Exec(fmt.Sprintf(`DROP TABLE IF EXISTS %s`, table)); err != nil {
		t.Logf("drop table %s: %v", table, err)
	}
}

func pgRow(id, requestID string) Row {
	return Row{
		ID:           id,
		OriginID:     id,
		MonotonicSeq: 1,
		Timestamp:    time.Now().UTC(),
		Code:         "USER_CREATED",
		Action:       "create",
		Severity:     SevInfo,
		ServiceID:    "svc-pg",
		SourceNodeID: "node-1",
		Actor:        "alice",
		ActorKind:    "user",
		Target:       "u-42",
		Category:     "account",
		Domain:       "user",
		Method:       "sdk",
		RequestID:    requestID,
		Detail:       map[string]any{"env": "test"},
	}
}

func TestPostgres_NewStore_InvalidTableName(t *testing.T) {
	db := pgDB(t)
	_, err := NewPostgresStore(db, "bad-name!")
	if err == nil {
		t.Fatal("expected error for invalid table name")
	}
}

func TestPostgres_InsertAndQuery(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_insert")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}

	row := pgRow("evt-pg-001", "reqaabbccdd11")
	if err := store.Insert(context.Background(), row); err != nil {
		t.Fatalf("Insert: %v", err)
	}

	got, err := store.Query(context.Background(), Filter{RequestID: "reqaabbccdd11"})
	if err != nil {
		t.Fatalf("Query: %v", err)
	}
	if len(got) != 1 {
		t.Fatalf("expected 1 row, got %d", len(got))
	}
	if got[0].ID != "evt-pg-001" {
		t.Errorf("id = %q, want evt-pg-001", got[0].ID)
	}
	if got[0].Code != "USER_CREATED" {
		t.Errorf("code = %q, want USER_CREATED", got[0].Code)
	}
	if got[0].ServiceID != "svc-pg" {
		t.Errorf("service_id = %q, want svc-pg", got[0].ServiceID)
	}
}

func TestPostgres_InsertIdempotent(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_idempotent")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}

	row := pgRow("evt-pg-dup", "reqdup000001")
	if err := store.Insert(context.Background(), row); err != nil {
		t.Fatalf("first Insert: %v", err)
	}
	if err := store.Insert(context.Background(), row); err != nil {
		t.Fatalf("second Insert (idempotent): %v", err)
	}

	got, err := store.Query(context.Background(), Filter{RequestID: "reqdup000001"})
	if err != nil {
		t.Fatalf("Query: %v", err)
	}
	if len(got) != 1 {
		t.Fatalf("expected 1 row after duplicate insert, got %d", len(got))
	}
}

func TestPostgres_ListUnshipped(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_unshipped")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}

	for i := 0; i < 3; i++ {
		r := pgRow(fmt.Sprintf("evt-unshipped-%d", i), fmt.Sprintf("req%012d", i))
		r.MonotonicSeq = int64(i + 1)
		if err := store.Insert(context.Background(), r); err != nil {
			t.Fatalf("Insert %d: %v", i, err)
		}
	}

	rows, err := store.ListUnshipped(context.Background(), 10)
	if err != nil {
		t.Fatalf("ListUnshipped: %v", err)
	}
	if len(rows) != 3 {
		t.Fatalf("expected 3 unshipped rows, got %d", len(rows))
	}
	for _, r := range rows {
		if r.ShippedAt != nil {
			t.Errorf("row %s should not have shipped_at set", r.ID)
		}
	}
}

func TestPostgres_MarkShipped(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_shipped")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}

	ids := []string{"evt-ship-1", "evt-ship-2", "evt-ship-3"}
	for i, id := range ids {
		r := pgRow(id, fmt.Sprintf("reqship%06d", i))
		r.MonotonicSeq = int64(i + 1)
		if err := store.Insert(context.Background(), r); err != nil {
			t.Fatalf("Insert %s: %v", id, err)
		}
	}

	if err := store.MarkShipped(context.Background(), ids[:2]); err != nil {
		t.Fatalf("MarkShipped: %v", err)
	}

	unshipped, err := store.ListUnshipped(context.Background(), 10)
	if err != nil {
		t.Fatalf("ListUnshipped: %v", err)
	}
	if len(unshipped) != 1 {
		t.Fatalf("expected 1 unshipped row, got %d", len(unshipped))
	}
	if unshipped[0].ID != "evt-ship-3" {
		t.Errorf("expected evt-ship-3 unshipped, got %s", unshipped[0].ID)
	}
}

func TestPostgres_MarkShipped_Empty(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_ship_empty")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}
	if err := store.MarkShipped(context.Background(), nil); err != nil {
		t.Fatalf("MarkShipped(nil): %v", err)
	}
}

func TestPostgres_Purge(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_purge")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}

	old := pgRow("evt-old", "reqold0000001")
	old.Timestamp = time.Now().UTC().Add(-48 * time.Hour)
	old.MonotonicSeq = 1
	if err := store.Insert(context.Background(), old); err != nil {
		t.Fatalf("Insert old: %v", err)
	}

	recent := pgRow("evt-recent", "reqrecent0001")
	recent.MonotonicSeq = 2
	if err := store.Insert(context.Background(), recent); err != nil {
		t.Fatalf("Insert recent: %v", err)
	}

	cutoff := time.Now().UTC().Add(-24 * time.Hour)
	n, err := store.Purge(context.Background(), cutoff, false)
	if err != nil {
		t.Fatalf("Purge: %v", err)
	}
	if n != 1 {
		t.Fatalf("expected 1 purged row, got %d", n)
	}

	rows, err := store.Query(context.Background(), Filter{})
	if err != nil {
		t.Fatalf("Query after purge: %v", err)
	}
	if len(rows) != 1 || rows[0].ID != "evt-recent" {
		t.Fatalf("expected only evt-recent to remain, got %v", rows)
	}
}

func TestPostgres_Purge_RespectUnshipped(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_purge_unshipped")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}

	shipped := pgRow("evt-shipped", "reqshipped001")
	shipped.Timestamp = time.Now().UTC().Add(-48 * time.Hour)
	shipped.MonotonicSeq = 1
	if err := store.Insert(context.Background(), shipped); err != nil {
		t.Fatalf("Insert shipped: %v", err)
	}
	if err := store.MarkShipped(context.Background(), []string{"evt-shipped"}); err != nil {
		t.Fatalf("MarkShipped: %v", err)
	}

	unshipped := pgRow("evt-unshipped", "requnsent0001")
	unshipped.Timestamp = time.Now().UTC().Add(-48 * time.Hour)
	unshipped.MonotonicSeq = 2
	if err := store.Insert(context.Background(), unshipped); err != nil {
		t.Fatalf("Insert unshipped: %v", err)
	}

	cutoff := time.Now().UTC().Add(-24 * time.Hour)
	n, err := store.Purge(context.Background(), cutoff, true)
	if err != nil {
		t.Fatalf("Purge(respectUnshipped=true): %v", err)
	}
	if n != 1 {
		t.Fatalf("expected 1 purged row, got %d", n)
	}

	rows, err := store.ListUnshipped(context.Background(), 10)
	if err != nil {
		t.Fatalf("ListUnshipped: %v", err)
	}
	if len(rows) != 1 || rows[0].ID != "evt-unshipped" {
		t.Fatalf("unshipped row should survive respectUnshipped purge, got %v", rows)
	}
}

func TestPostgres_QueryFilters(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_filters")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}

	r1 := pgRow("evt-filter-1", "reqfilter00001")
	r1.MonotonicSeq = 1
	r1.Code = "USER_CREATED"
	r1.Domain = "user"

	r2 := pgRow("evt-filter-2", "reqfilter00002")
	r2.MonotonicSeq = 2
	r2.Code = "USER_DELETED"
	r2.Domain = "user"

	for _, r := range []Row{r1, r2} {
		if err := store.Insert(context.Background(), r); err != nil {
			t.Fatalf("Insert %s: %v", r.ID, err)
		}
	}

	got, err := store.Query(context.Background(), Filter{Code: "USER_CREATED"})
	if err != nil {
		t.Fatalf("Query by Code: %v", err)
	}
	if len(got) != 1 || got[0].ID != "evt-filter-1" {
		t.Fatalf("expected evt-filter-1, got %v", got)
	}

	got, err = store.Query(context.Background(), Filter{Domain: "user"})
	if err != nil {
		t.Fatalf("Query by Domain: %v", err)
	}
	if len(got) != 2 {
		t.Fatalf("expected 2 rows for domain=user, got %d", len(got))
	}

	got, err = store.Query(context.Background(), Filter{SourceNodeID: "node-1"})
	if err != nil {
		t.Fatalf("Query by SourceNodeID: %v", err)
	}
	if len(got) != 2 {
		t.Fatalf("expected 2 rows for source_node_id=node-1, got %d", len(got))
	}
}

func TestPostgres_SchemaQualifiedTable(t *testing.T) {
	db := pgDB(t)

	schema := fmt.Sprintf("fasten_test_%d", time.Now().UnixNano())
	table := schema + ".audit"
	defer func() {
		db.Exec(fmt.Sprintf(`DROP SCHEMA IF EXISTS %s CASCADE`, schema))
	}()

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore with schema: %v", err)
	}

	row := pgRow("evt-schema-1", "reqschema00001")
	if err := store.Insert(context.Background(), row); err != nil {
		t.Fatalf("Insert: %v", err)
	}

	got, err := store.Query(context.Background(), Filter{RequestID: "reqschema00001"})
	if err != nil {
		t.Fatalf("Query: %v", err)
	}
	if len(got) != 1 || got[0].ID != "evt-schema-1" {
		t.Fatalf("expected evt-schema-1, got %v", got)
	}
}

func TestPostgres_PiiInDetail(t *testing.T) {
	db := pgDB(t)
	table := uniqueTable("pg_pii")
	defer dropTable(t, db, table)

	store, err := NewPostgresStore(db, table)
	if err != nil {
		t.Fatalf("NewPostgresStore: %v", err)
	}

	row := pgRow("evt-pii-1", "reqpii0000001")
	row.PiiInDetail = true
	row.Detail = map[string]any{"_redacted": "***", "_pii_in_detail": true}
	if err := store.Insert(context.Background(), row); err != nil {
		t.Fatalf("Insert: %v", err)
	}

	got, err := store.Query(context.Background(), Filter{RequestID: "reqpii0000001"})
	if err != nil {
		t.Fatalf("Query: %v", err)
	}
	if len(got) != 1 || !got[0].PiiInDetail {
		t.Fatalf("expected PiiInDetail=true, got %+v", got)
	}
}

File: js/test/store_postgres.test.mjs
/**
 * Integration tests for PostgresStore.
 *
 * Set FASTEN_TEST_POSTGRES_DSN to run against a real Postgres instance:
 *
 *   FASTEN_TEST_POSTGRES_DSN=postgresql://fasten:fasten@localhost:5432/fasten_test \
 *     node --test test/store_postgres.test.mjs
 *
 * All tests are skipped when the env-var is absent.
 */
import { test, after } from 'node:test';
import assert from 'node:assert/strict';
import { randomUUID } from 'node:crypto';
import { PostgresStore } from '../src/store/postgres.js';

const DSN = process.env.FASTEN_TEST_POSTGRES_DSN;
const skip = !DSN;

// One unique table for the entire test run — dropped in `after`.
const TABLE = `t_${randomUUID().replace(/-/g, '').slice(0, 12)}`;

let store = skip ? null : new PostgresStore(DSN, { table: TABLE });

// ── Helpers ────────────────────────────────────────────────────────────────

function _id() {
    return `evt-${randomUUID().replace(/-/g, '').slice(0, 20)}`;
}

let _seq = 0;
function makeRow(overrides = {}) {
    _seq++;
    return {
        id: _id(),
        origin_id: _id(),
        monotonic_seq: _seq,
        timestamp: new Date().toISOString(),
        code: 'USER_CREATED',
        action: 'create',
        severity: 'info',
        service_id: 'svc',
        source_node_id: 'node-1',
        tenant_id: null,
        actor: 'system',
        actor_kind: 'service',
        target: 'u-1',
        category: 'account',
        domain: 'user',
        method: 'sdk',
        request_id: randomUUID().replace(/-/g, '').slice(0, 12),
        detail: { email: 'a@b.com' },
        pii_in_detail: false,
        shipped_at: null,
        ...overrides,
    };
}

// ── Cleanup ────────────────────────────────────────────────────────────────

after(async () => {
    if (skip || !store) return;
    // Drop the shared table then close the pool.
    // Use a fresh pool because store._pool may be ended by close().
    const { default: pg } = await import('pg');
    const { Pool } = pg;
    const cleanup = new Pool({ connectionString: DSN });
    await cleanup.query(`DROP TABLE IF EXISTS ${TABLE}`);
    await cleanup.end();
    await store.close();
});

// ── Tests ──────────────────────────────────────────────────────────────────

// 1. insert + query by requestId
test('insert and query by requestId', { skip }, async () => {
    const rid = randomUUID().replace(/-/g, '').slice(0, 12);
    const row = makeRow({ request_id: rid });
    await store.insert(row);
    const results = await store.query({ requestId: rid });
    assert.equal(results.length, 1);
    assert.equal(results[0].id, row.id);
    assert.equal(results[0].code, 'USER_CREATED');
});

// 2. insert idempotent (ON CONFLICT DO NOTHING)
test('insert is idempotent — ON CONFLICT DO NOTHING', { skip }, async () => {
    const row = makeRow();
    await store.insert(row);
    await store.insert(row);
    const n = await store.count({ requestId: row.request_id });
    assert.equal(n, 1);
});

// 3. listUnshipped
test('listUnshipped returns rows with no shipped_at', { skip }, async () => {
    const row = makeRow();
    await store.insert(row);
    const unshipped = await store.listUnshipped(100);
    const ids = unshipped.map((r) => r.id);
    assert.ok(ids.includes(row.id), `expected ${row.id} in unshipped list`);
    for (const r of unshipped) {
        assert.equal(r.shippedAt, null, `row ${r.id} should have shippedAt=null`);
    }
});

// 4. markShipped removes from unshipped
test('markShipped removes row from listUnshipped', { skip }, async () => {
    const row = makeRow();
    await store.insert(row);

    // Confirm it starts unshipped.
    const before = await store.listUnshipped(500);
    assert.ok(before.some((r) => r.id === row.id));

    await store.markShipped([row.id]);

    const after = await store.listUnshipped(500);
    assert.ok(!after.some((r) => r.id === row.id),
        `${row.id} should no longer be in unshipped list`);

    // shipped_at should now be set.
    const [back] = await store.query({ requestId: row.request_id });
    assert.ok(back.shippedAt !== null, 'shippedAt should be set after markShipped');
});

// 5. purge respects unshipped
test('purge respects unshipped rows', { skip }, async () => {
    // Insert a row with an old timestamp but leave it unshipped.
    const oldTs = new Date(Date.now() - 31 * 24 * 60 * 60 * 1000).toISOString();
    const row = makeRow({ timestamp: oldTs });
    await store.insert(row);

    const before = await store.count({ requestId: row.request_id });
    assert.equal(before, 1);

    const deleted = await store.purge({ before: new Date().toISOString(), respectUnshipped: true });
    // Row is unshipped so it must NOT be deleted.
    assert.equal(deleted, 0, 'unshipped rows must not be purged');

    const after = await store.count({ requestId: row.request_id });
    assert.equal(after, 1);
});

// 6. purge shipped
test('purge deletes shipped rows older than cutoff', { skip }, async () => {
    const oldTs = new Date(Date.now() - 31 * 24 * 60 * 60 * 1000).toISOString();
    const row = makeRow({ timestamp: oldTs });
    await store.insert(row);
    await store.markShipped([row.id]);

    const deleted = await store.purge({ before: new Date().toISOString(), respectUnshipped: true });
    assert.ok(deleted >= 1, `expected at least 1 deleted row; got ${deleted}`);

    const remaining = await store.count({ requestId: row.request_id });
    assert.equal(remaining, 0);
});

// 7. count filtered
test('count filtered by code', { skip }, async () => {
    const rid = randomUUID().replace(/-/g, '').slice(0, 12);
    await store.insert(makeRow({ code: 'USER_CREATED', request_id: rid }));
    await store.insert(makeRow({ code: 'USER_CREATED', request_id: rid }));
    await store.insert(makeRow({ code: 'USER_DELETED', request_id: rid }));

    const total = await store.count({ requestId: rid });
    assert.equal(total, 3);

    const created = await store.count({ requestId: rid, code: 'USER_CREATED' });
    assert.equal(created, 2);

    const deleted = await store.count({ requestId: rid, code: 'USER_DELETED' });
    assert.equal(deleted, 1);
});

// 8. round-trip detail JSON
test('round-trip detail JSON', { skip }, async () => {
    const detail = { nested: { k: [1, 2, 3] }, flag: true, n: null };
    const row = makeRow({ detail });
    await store.insert(row);
    const [back] = await store.query({ requestId: row.request_id });
    assert.deepEqual(back.detail, detail);
});

// 9. piiInDetail round-trips
test('piiInDetail round-trips as boolean', { skip }, async () => {
    const row = makeRow({ pii_in_detail: true });
    await store.insert(row);
    const [back] = await store.query({ requestId: row.request_id });
    assert.equal(back.piiInDetail, true);
});

// 10. schema-qualified table
test('schema-qualified table (fasten.<TABLE>)', { skip }, async () => {
    const schemaTable = `fasten.${TABLE}_sq`;
    const schemaStore = new PostgresStore(DSN, { table: schemaTable });

    try {
        const row = makeRow();
        await schemaStore.insert(row);

        const count = await schemaStore.count();
        assert.equal(count, 1);

        const [back] = await schemaStore.query({ requestId: row.request_id });
        assert.equal(back.id, row.id);
    } finally {
        // Cleanup: drop table + schema.
        const { default: pg } = await import('pg');
        const { Pool } = pg;
        const cleanup = new Pool({ connectionString: DSN });
        await cleanup.query(`DROP TABLE IF EXISTS ${schemaTable}`);
        await cleanup.query('DROP SCHEMA IF EXISTS fasten CASCADE');
        await cleanup.end();
        await schemaStore.close();
    }
});
